File: trapRainWater.cpp
// Time Complexity :O(n) where n in the number elements in the vector
// Space Complexity : O(1)  
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
public:
    int trap(vector<int>& height) {
        int left = 0;
        int right = height.size()-1;
        int rWall = 0;
        int lWall = 0;
        int trap = 0;
        while(left <= right){
            if(lWall <=rWall){
                if(lWall < height[left]) lWall = height[left];
                else trap+= 1 * (lWall - height[left]);
                left++;
            }
            else{
                if(rWall < height[right]) rWall = height[right];
                else trap+= 1 * (rWall-height[right]);
                right--;
            }
        }
        return trap;
    }
};

File: TrapRainWater.cs

// Time Complexity : O(n))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
	// 1) lw is height of left wall
	// 2) l is left pinter
	// 3) rw is height of right wall
	// 4) r is right pointe
	// 5) Assuming right of left wall , there will be some bigger right wall, where we can trap the water
	// 	a. RW (height of r) > LW (height of l) -> then we can trap the water
	// 6) We have wall on right side, we will handle separately.


public int Trap(int[] height) {
     
    if(height == null || height.Length == 0)
        return 0;
    
    int result = 0;
    int lw = 0, rw = 0;
    int l = 0, r = height.Length - 1;
    
    while(l <= r)
    {
        if(lw <= rw)
        {
            if(lw >= height[l])
                result += lw - height[l];
            else
                lw = height[l];
            l++;
        }
        else
        {
            if(rw >= height[r])
                result += rw - height[r];
            else
                rw = height[r];
            r--;
        }
    }
    
    return result;
}